src/lib/router.ts

```diff
@@ -74,119 +67,190 @@ const allRoutes: RouteDefinition[] = [
-
-    return route
-  }, [])
-}
-
-interface RouteDefinition {
+type RouteDefinition = {
   label: string
   name: string
   path?: string
   routes?: RouteDefinition[]
 }
 
-interface Route {
+export class Route {
   label: string
   name: string
-  fullName: string
   path: string
-  fullPath: string
-  routes: Routes
-}
 
-type Routes = Route[]
+  private _activePath: string
+  private _parent: Route
+  private _routes: Route[]
 
-export class Router {
-  private _flattenedRoutes: Routes
+  constructor({
+    label,
+    name,
+    path,
+    routes,
+    parent,
+  }: {
+    label: string
+    name: string
+    path: string
+    routes?: Route[]
+    parent?: Route
+  }) {
+    this.label = label
+    this.name = name
+    this.path = path
 
-  activePath: String
-  routes: Route[]
-  rootRouter: Router
+    this.routes = routes ? routes : []
 
-  constructor(routes?: Route[], rootRouter?: Router) {
-    if (routes) {
-      this.routes = routes
-    } else {
-      this.routes = transformRoutes(allRoutes)
+    if (parent) {
+      this.parent = parent
     }
+  }
+
+  get fullName(): string {
+    return [this.parent && this.parent.fullName, this.name]
+      .filter(part => part && part !== "")
+      .join(".")
+  }
+
+  get fullPath(): string {
+    return [this.parent && this.parent.fullPath, this.path]
+      .filter(part => part && part !== "")
+      .join("")
+  }
+
+  get parent(): Route {
+    return this._parent
+  }
 
-    if (rootRouter) {
-      this.rootRouter = rootRouter
+  set parent(parent: Route) {
+    this._parent = parent
+    if (!parent.routes.includes(this)) {
+      parent.routes.push(this)
     }
   }
 
-  // Flatten all routes
-  get flattenedRoutes(): Routes {
-    if (!this._flattenedRoutes) {
-      function flatten(routes: Route[]) {
-        return routes.reduce((flattenedRoutes, { routes, ...rest }) => {
-          return [
-            ...flattenedRoutes,
-            ...(routes.length > 0 ? flatten(routes) : rest),
-          ]
-        }, [])
+  get routes(): Route[] {
+    return this._routes
+  }
+
+  set routes(routes: Route[]) {
+    // if we're ever going to remove routes we should address that here
+    this._routes = routes
+    routes.forEach(route => {
+      if (route.parent && route.parent !== this) {
+        throw new Error(
+          `Cannot add ${route.fullName} to ${this.fullName}, because it already belongs to ${route.parent.fullName}`
+        )
       }
+      route.parent = this
+    })
+  }
 
-      this._flattenedRoutes = flatten(this.routes)
+  get allRoutes(): Route[] {
+    let flatten = function(routes: Route[]) {
+      return routes.reduce((result, route) => {
+        return [...result, ...[route], ...flatten(route.routes)]
+      }, [])
     }
 
-    return this._flattenedRoutes
+    return flatten(this.routes)
   }
 
-  // Return the active route
-  get activeRoute(): Route {
-    if (this.rootRouter) {
-      return this.rootRouter.activeRoute
+  get pages(): Route[] {
```

Comment:
`page`, great name 👍 Maybe make this `allPages`?

Does this mean we could get rid of `allRoutes`?

Reply:
👍 

Still need all routes so we can find sub routers by path `router.routerFor('/something/something')` where `something.something` is not a page.

Reply:
interesting. Maybe `routes` should be private... we're using `router.routerFor(...)` to get scoped pages right? Maybe it should be `router.pagesUnder('/something/')`.

Maybe there's another data structure that's just about ordered pages.